File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int r = grid.length;
        int c = grid[0].length;
        int count = 0;
        HashSet<Integer> set = new HashSet<>();
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == '1') {
                    if (dfs(i, j, r, c, set, grid)) {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    private boolean dfs(int i, int j, int r, int c, HashSet<Integer> set, char[][] grid) {
        if (i >= r || j >= c || i < 0 || j < 0 || set.contains(i * c + j) || grid[i][j] == '0') {
            return false;
        }
        set.add(i * c + j);
        dfs(i + 1, j, r, c, set, grid);
        dfs(i - 1, j, r, c, set, grid);
        dfs(i, j + 1, r, c, set, grid);
        dfs(i, j - 1, r, c, set, grid);
        return true;
    }
}

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        v = set()
        count =0


        def dfs(r,c):
            if r<0 or r >= rows or c <0 or c>=cols or (r,c) in v or grid[r][c] == "0":
                return
            v.add((r,c))
            dfs(r+1,c)
            dfs(r-1,c)
            dfs(r,c+1)
            dfs(r,c-1)




        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == "1" and (i,j) not in v:
                    count+=1
                    dfs(i,j)
        return count